### PR TITLE
Move to a single cache for all responses

### LINKED ISSUE #86: Possible to use a single cache?
Chatting with one of the engineers and the topic of using multiple caches (i.e. one per asset) could be poor in terms of the number of instances of caches in memory at any one time.

Is there a specific reason precache couldn't use a single cache?

### PR DESCRIPTION
**Pull request by jeffposnick**
R: @addyosmani @gauntface @wibblymat 

I think I'm happier with the code using this approach, as it also simplifies some of the logic by using `Map` and `Set`, neither of which were widely available when the code was originally written.

The one drawback that I can see is that it's no longer possible to use `caches.match()` to pick up a given `Response` from arbitrary code that the user writes, since the `Request` objects used as cache keys now have URL parameters appended to them. That's not an advertised use case, though, so I'm not worried.

This will be part of a forthcoming 4.0 release, so I'm not ready to update the `package.json` metadata yet.

Closes #86

### CODE REVIEW

File: lib/sw-precache.js [hunk 1]
```diff
@@ -181,7 +181,7 @@ function generate(params, callback) {
     var runtimeCaching;
     var swToolboxCode;
     if (params.runtimeCaching) {
-      var pathToSWToolbox = require.resolve('sw-toolbox/sw-toolbox.js');
+      var pathToSWToolbox = require.resolve('sw-toolbox/build/sw-toolbox.js');
```
Comment (wibblymat) [inline lib/sw-precache.js]: Ack, my bad. I read through the list of commits to decide what kind of version bump I needed for the latest release, and missed that this caused a breaking change.

File: service-worker.tmpl [hunk 2]
```diff
@@ -34,10 +34,8 @@
 importScripts(<%= importScripts %>);
 <% } %>
 
-/* eslint-disable quotes, comma-spacing */
 var PrecacheConfig = <%= precacheConfig %>;
-/* eslint-enable quotes, comma-spacing */
-var CacheNamePrefix = 'sw-precache-<%= version %>-<%= cacheId %>-' + (self.registration ? self.registration.scope : '') + '-';
+var CacheName = 'sw-precache-<%= version %>-<%= cacheId %>-' + (self.registration ? self.registration.scope : '');
```
Comment (wibblymat) [inline service-worker.tmpl]: Google's style guide says that starting with a capital letter is reserved for classes (well, constructors).

Comment (wibblymat) [inline service-worker.tmpl]: Oh, I see that's not being followed throughout the file actually, so I guess this is actually fitting in with the surrounding code.

File: service-worker.tmpl [hunk 3]
```diff
@@ -47,141 +45,108 @@ var IgnoreUrlParametersMatching = [<%= ignoreUrlParametersMatching %>];
...
+  PrecacheConfig.map(function(item) {
+    var relativeUrl = item[0];
+    var hash = item[1];
+    var absoluteUrl = new URL(relativeUrl, self.location);
+    var cacheKey = createCacheKey(absoluteUrl, HashParamName, hash);
+    return [absoluteUrl.toString(), cacheKey];
+  })
+);
+
+function setOfCachedUrls(cache) {
```
Comment (wibblymat) [inline service-worker.tmpl]: Could ES6-ify new code as it comes in even if you aren't going to redo existing code.

Using this function as an example:
```
function setOfCacheUrls(cache) {
  return cache.keys()
    .then(requests => requests.map(req => req.url))
    .then(urls => new Set(urls));
}
```

**Comment by wibblymat** [comment]
This is great!

**Comment by wibblymat** [comment]
Regarding the point about caches.match... could the parameters be stored on the *Response* instead? So, construct the modified URL, fetch that, and `cache.put` with the original Request as the key.

I guess it makes it harder to look this stuff up...

File: service-worker.tmpl (hunk 2 above)
Comment (jeffposnick) [inline service-worker.tmpl]: It was a global scope convention, but I'm happy to change it to lowercase.

File: service-worker.tmpl (hunk 3 above)
Comment (jeffposnick) [inline service-worker.tmpl]: I'd prefer to keep things consistent throughout, and move to ES2015 _en masse_. https://github.com/GoogleChrome/sw-precache/issues/82

**Comment by jeffposnick** [comment]
That's an interesting idea re: using `Response` objects—maybe in a header?—to store the cache versioning info. The drawbacks I see there are:

- I'd have to do a bunch of `cache.match()` operations to get all the versioning info within the `install` handler, vs. (what I'm assuming is a cheaper?) `cache.keys()`.
- Most of the fields in a `Response` object returned by `fetch()` are [read-only](https://fetch.spec.whatwg.org/#response-class), so I'd end up constructing new `Response` objects and synthesizing a combination of the original and an updated headers in order to do that.
- I wonder if there are CORS concerns from the perspective of the controlled page if it ends up using a `Response` that has extra headers set?

Any thoughts on those?

File: lib/sw-precache.js (hunk 1 above)
Comment (addyosmani) [inline lib/sw-precache.js]: Honest oversight :) IMO we should revert this change upstream (or revert, retag, intro as a major change). It's starting to cause issues to be filed on projects elsewhere (Polymer CLI etc)

File: service-worker.tmpl (hunk 3 above)
Comment (addyosmani) [inline service-worker.tmpl]: I agree with moving to the newer syntax in one go given you have a plan for it already.

**Comment by addyosmani** [comment]
LGTM. My only real comment was to do with the toolbox dep. Everything else appears in order here.

File: lib/sw-precache.js (hunk 1 above)
Comment (gauntface) [inline lib/sw-precache.js]: sw-toolbox could be changed to copy the build back to the root of the project for the release. This is the result of *not* having the built sw-toolbox in the master branch.

**Comment by jeffposnick** [comment]
I'm going to go ahead with the strategy of using a custom search param in the `Request` URL to version each resource. Down the line, if https://github.com/slightlyoff/ServiceWorker/issues/854 is implemented, it could make sense to switch to a URL fragment.
